src/modules.adoc

```diff
@@ -0,0 +1,361 @@
+unbound reference.
+Otherwise, the name or address is resolved within that module’s exported macro array.
+
+NOTE: An unqualified macro name can change meaning in the middle of a module: it could be
+imported and used with that meaning, then a declaration shadows that name and gives it a new
+meaning.
+
+=== Module Versioning
+
+Every module definition has a _spec version_ that gives the definition its meaning in terms of
+acceptable syntax, available features, and so on.
+A module’s spec version is expressed in terms of a specific Ion version; the meaning of the
+module is as defined by that version of the Ion specification.
+
+The spec version of a shared or tunneled module must be declared explicitly using an annotation
+of the form `$ion_1___N__`.
+This allows the module to be serialized using any version of Ion, and its meaning will not change.
+
+[{nrm}]
+----
+$ion_shared_module::$ion_1_1::(
+  (*catalog_key* "com.example.symtab" 3)
+  (*symbol_table* ...)
+  (*macro_table* ...)
+)
+
+$ion_shared_symbol_table::{
+  name: "com.example.symtab", version: 3,
+  symbols: […],
+  module: $ion_1_1::(      // Spec version is 1.1
+    // Semantics of this module are specified by Ion 1.1, regardless of the
+    // enclosing document's Ion version.
+    ...
+  )
+}
+----
+
+The spec version of an inline module is always the same as the Ion version of its enclosing
+segment.
+
+[{nrm}]
+----
+*$ion_1_1*
+$ion_encoding::(
+  (*module* M1 ...)    // Module semantics specified by Ion 1.1
+  ...
+}
+...
+$ion_1_3
+$ion_encoding::(
+  (*module* M2 ...)    // Module semantics specified by Ion 1.3
+  ...
+}
+...                  // Assuming no IVM
+$ion_encoding::(
+  (*module* M3 ...)    // Module semantics specified by Ion 1.3
+  ...
+}
+----
+
+To ensure that all consumers of a module can properly understand it, a module can only import
+shared modules defined with the same or earlier spec version.
+
+=== Inline, Shared, and Tunneled Modules
+
+Inline modules are defined within an `$ion_encoding` directive, and are available only within the
+enclosing document.
+Their scope is lexical; they can be used immediately following their definition, up until the
+next directive, at which point they’ll either be retained by the new encoding environment, or
+made unavailable.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=inline-module]
+|===
+
+Inline modules always have a symbolic name given at the point of definition.
+They inherit their spec version from the surrounding document, and they have no content version.
+
+Shared modules exist independently of the documents that use them.
+They are identified by a _catalog key_ consisting of a string name and an integer version.
+When consumed by a document or another module, they are given a local identifier.
+
+[{bnf}]
+|===
+include::grammar.adoc[tag=shared-module]
+|===
+
+Tunneled modules are shared modules that are defined within a shared symbol table definition.
```

Comment:
This probably isn't the place for it, but somewhere we'll need a disclaimer saying this is mostly for backwards compatibility and new modules should be encoded with the modern syntax.